test/e2e/pkg/infra/infra.go

```diff
@@ -0,0 +1,84 @@
+package infra
+
+import (
+	"context"
+
+	e2e "github.com/tendermint/tendermint/test/e2e/pkg"
+)
+
+// TestnetInfra provides an API for manipulating the infrastructure of a
+// specific testnet.
+type TestnetInfra interface {
+	//
+	// Overarching testnet infrastructure management.
+	//
+
+	// Setup generates any necessary configuration for the infrastructure
+	// provider during testnet setup.
+	Setup(ctx context.Context) error
+
+	// Stop will stop all running processes throughout the testnet without
+	// destroying any infrastructure.
+	Stop(ctx context.Context) error
+
+	// Pause will pause all processes in the testnet.
+	Pause(ctx context.Context) error
+
+	// Unpause will resume a paused testnet.
+	Unpause(ctx context.Context) error
+
+	// ShowLogs prints all logs for the whole testnet to stdout.
+	ShowLogs(ctx context.Context) error
+
+	// TailLogs tails the logs for all nodes in the testnet, if this is
+	// supported by the infrastructure provider.
+	TailLogs(ctx context.Context) error
+
+	// Cleanup stops and destroys all running testnet infrastructure and
+	// deletes any generated files.
+	Cleanup(ctx context.Context) error
+
+	//
+	// Node management, including node infrastructure.
+	//
+
+	// StartNode provisions infrastructure for the given node and starts it.
+	StartNode(ctx context.Context, node *e2e.Node) error
+
+	// DisconnectNode modifies the specified node's network configuration such
+	// that it becomes bidirectionally disconnected from the network (it cannot
+	// see other nodes, and other nodes cannot see it).
+	DisconnectNode(ctx context.Context, node *e2e.Node) error
+
+	// ConnectNode modifies the specified node's network configuration such
+	// that it can become bidirectionally connected.
+	ConnectNode(ctx context.Context, node *e2e.Node) error
```

Comment:
In docker, this works by adding the running container to the docker network. This effectively connects it to the other running nodes. The container can be configured to start on the same network as the other nodes, but if it is not then calling connect will put the nodes onto the same network.

https://docs.docker.com/engine/reference/commandline/network_connect/

Reply:
So this can be used to simulate the network itself failing?

How would this be achieved in DO?

Reply:
The easiest way is to [modify the node's iptables](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/iptables_module.html), cutting off incoming and outgoing connections to/from certain ports to "disconnect" a node and then restoring them to "connect" it.

For more advanced network outage simulation we'd need to install something like [Toxiproxy](https://github.com/shopify/toxiproxy) on the nodes and configure it with the networking parameters we'd want.

Reply:
> So this can be used to simulate the network itself failing?

Correct. 

> How would this be achieved in DO?

A few ways I can think of
1. Digital Ocean has a concept of a [firewall](https://registry.terraform.io/providers/digitalocean/digitalocean/latest/docs/resources/firewall) that can be used to isolate nodes from IP ranges. I haven't looked much into it, but it looks like it can do the trick.
2. If the firewall rules are too cumbersome or don't give us the flexibility we need, the backup is always to use [IPtables](https://linux.die.net/man/8/iptables) rules `DROP` or `REJECT` rules depending on what we want to achieve. (Drop simulates a slow app not able to keep up with its ACKs, REJECT simulates a node that is completely unreachable behind a firewall (sends RST)).

Digital Ocean does seem to have an idea of a VPC (virtual private cloud networking) that allows nodes to all be on a shared network, but adding/removing nodes to the VPC seems sketchy and bad. 
https://docs.digitalocean.com/products/networking/vpc/how-to/migrate-resources/